pystiche/papers/sanakoyeu_et_al_2018/core.py

```diff
@@ -0,0 +1,302 @@
+    num_epochs: Optional[int] = None,
+    get_optimizer: Optional[sanakoyeu_et_al_2018_optimizer] = None,
+):
+    if device is None:
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    if transformer is None:
+        transformer = SanakoyeuEtAl2018Transformer()
+        transformer = transformer.train()
+    transformer = transformer.to(device)
+
+    if discriminator_operator is None:
+        discriminator_operator = sanakoyeu_et_al_2018_discriminator_operator()
+
+    if discriminator_criterion is None:
+        discriminator_criterion = SanakoyeuEtAl2018DiscriminatorLoss(
+            discriminator_operator
+        )
+        discriminator_criterion = discriminator_criterion.eval()
+    discriminator_criterion = discriminator_criterion.to(device)
+
+    if transformer_criterion is None:
+        transformer_criterion = sanakoyeu_et_al_2018_transformer_loss(
+            transformer.encoder,
+            impl_params=impl_params,
+            style_loss=discriminator_operator,
+        )
+        transformer_criterion = transformer_criterion.eval()
+    transformer_criterion = transformer_criterion.to(device)
+
+    if get_optimizer is None:
+        get_optimizer = sanakoyeu_et_al_2018_optimizer
+
+    def transformer_criterion_update_fn(content_image, criterion):
+        criterion.set_content_image(content_image)
+
+    if impl_params:
+        return gan_optim_loop(
+            content_image_loader,
+            style_image_loader,
+            transformer,
+            discriminator_operator,
+            discriminator_criterion,
+            transformer_criterion,
+            transformer_criterion_update_fn,
+            get_optimizer=get_optimizer,
+            impl_params=impl_params,
+            device=device,
+        )
+    else:
+        if discriminator_lr_scheduler is None:
+            discriminator_optimizer = get_optimizer(
+                discriminator_operator.get_discriminator_parameters()
+            )
+            discriminator_lr_scheduler = sanakoyeu_et_al_2018_lr_scheduler(
+                discriminator_optimizer
+            )
+
+        if transformer_lr_scheduler is None:
+            transformer_optimizer = get_optimizer(transformer)
+            transformer_lr_scheduler = sanakoyeu_et_al_2018_lr_scheduler(
+                transformer_optimizer
+            )
+
+        if num_epochs is None:
+            num_epochs = 3
+
+        return epoch_gan_optim_loop(
+            content_image_loader,
+            style_image_loader,
+            transformer,
+            discriminator_operator,
+            num_epochs,
+            discriminator_criterion,
+            transformer_criterion,
+            transformer_criterion_update_fn,
+            discriminator_lr_scheduler=discriminator_lr_scheduler,
+            transformer_lr_scheduler=transformer_lr_scheduler,
+            impl_params=impl_params,
+            device=device,
+        )
```

Comment:
Yes, we could merge them into one block, since the first call would basically be no different than setting `num_epochs=1`. However, wouldn't it be confusing to call `epoch_gan_optim_loop` with `num_epochs=1` and `lr_scheduler` which are not used? To show the difference between `impl_params` `True `and `False `I have deliberately divided it this way. Thoughts?